File: 2140-longest-subsequence-repeated-k-times/2140-longest-subsequence-repeated-k-times.java
class Solution {
    public String longestSubsequenceRepeatedK(String s, int k) {
        // TODO 1: Count frequency of each character
        int [] freq = new int[26];
        for(char c : s.toCharArray()){
            freq[c - 'a']++;
        }
        // TODO 2: Prepare list of valid characters (those with freq >= k), reverse-lex order
        List<Character> validChars = new ArrayList<>();
        for(int i = 25; i >= 0; i--){
            if(freq[i] >= k){
                validChars.add((char)(i + 'a'));
            }
        }
        // TODO 3: BFS initialization: start with empty string
        Queue<String> queue = new LinkedList<>();
        queue.offer("");
        String res = "";
        
        // TODO 4: BFS to build candidates
        while(!queue.isEmpty()){
            String curr = queue.poll();

            // TODO 5: Try appending each valid character
            for(char c : validChars){
                String next = curr + c;

                // TODO 6: Check if next * k is a subsequence of s
                if(isKRepeatedSubsequence(s, next, k)){
                    // TODO 7: If longer or lexicographically larger, update res
                    if(next.length() > res.length() ||
                        (next.length() == res.length() && next.compareTo(res) > 0)){
                            res = next;
                        }
                        // TODO 8: Add to queue to explore longer subsequences
                        queue.offer(next);
                }
            }
        }

        return res;
    }
    // TODO 9: Check if target * k is subsequence of source
    private boolean isKRepeatedSubsequence(String s, String t, int k){
        int tlen = t.length();
        int index = 0, count = 0;

        for(char c : s.toCharArray()){
            if(c == t.charAt(index)){
                index++;
                if(index == tlen){
                    count++;
                    if(count == k) return true;
                    index = 0;   // restart t again
                }
            }
        }
        return false;
    }
}

File: 2140-longest-subsequence-repeated-k-times/2140-longest-subsequence-repeated-k-times.cpp
class Solution {
public:
    string res = "";

    // Check if 'curr' repeated k times is a subsequence of 's'
    bool isSubsequence(string &s, string &curr, int k) {
        int i = 0, j = 0, count = 0;
        int n = s.length(), m = curr.length();

        while (i < n) {
            if (s[i] == curr[j]) {
                j++;
                if (j == m) {
                    count++;
                    j = 0;
                    if (count == k) return true;
                }
            }
            i++;
        }
        return false;
    }

    void backtracking(string &s, string &curr, vector<bool>& canUse, vector<int>& requiredFreq, int k, int maxLen) {
        if (curr.length() > maxLen) return;

        if (((curr.length() > res.length()) || (curr.length() == res.length() && curr > res)) && isSubsequence(s, curr, k)) {
            res = curr;
        }

        for (int i = 25; i >= 0; i--) { // start from 'z' for lexicographically larger
            if (canUse[i] && requiredFreq[i] > 0) {
                curr.push_back('a' + i);
                requiredFreq[i]--;
                backtracking(s, curr, canUse, requiredFreq, k, maxLen);
                requiredFreq[i]++;
                curr.pop_back();
            }
        }
    }

    string longestSubsequenceRepeatedK(string s, int k) {
        int n = s.length();
        int freq[26] = {};
        for (char &ch : s) {
            freq[ch - 'a']++;
        }

        vector<bool> canUse(26, false);
        vector<int> requiredFreq(26, 0);
        for (int i = 0; i < 26; i++) {
            if (freq[i] >= k) {
                canUse[i] = true;
                requiredFreq[i] = freq[i] / k; // at most this many times can be used
            }
        }

        int maxLen = n / k;
        string curr;
        backtracking(s, curr, canUse, requiredFreq, k, maxLen);
        return res;
    }
};
